src/main/kotlin/com/chattriggers/ctjs/minecraft/libs/renderer/Renderer3d.kt

```diff
@@ -0,0 +1,303 @@
+import net.minecraft.client.render.VertexFormat
+import org.joml.Vector3f
+import org.mozilla.javascript.NativeObject
+import java.awt.Color
+import kotlin.math.sqrt
+
+object Renderer3d {
+    private var firstVertex = true
+    private var began = false
+
+    private val tessellator = Tessellator.getInstance()
+    private val worldRenderer = UGraphics(tessellator.buffer)
+
+    /**
+     * Begin drawing with the world renderer
+     *
+     * @param drawMode the GL draw mode
+     * @param vertexFormat The [VertexFormat] to use for drawing
+     * @return the Renderer to allow for method chaining
+     * @see com.chattriggers.ctjs.minecraft.libs.renderer.Shape.setDrawMode
+     */
+    @JvmStatic
+    @JvmOverloads
+    fun begin(
+        drawMode: Renderer.DrawMode = Renderer.DrawMode.QUADS,
+        vertexFormat: Renderer.VertexFormat = Renderer.VertexFormat.POSITION,
+    ) = apply {
+        Renderer.pushMatrix()
+        Renderer.enableBlend()
+        Renderer.tryBlendFuncSeparate(770, 771, 1, 0)
+
+        worldRenderer.beginWithDefaultShader(drawMode.toUC(), vertexFormat.toMC())
+
+        firstVertex = true
+        began = true
+    }
+
+    /**
+     * Sets a new vertex in the Tessellator.
+     *
+     * @param x the x position
+     * @param y the y position
+     * @param z the z position
+     * @return the Tessellator to allow for method chaining
+     */
+    @JvmStatic
+    fun pos(x: Float, y: Float, z: Float) = pos(x.toDouble(), y.toDouble(), z.toDouble())
+
+    @JvmStatic
+    fun pos(x: Double, y: Double, z: Double) = apply {
+        if (!began)
+            begin()
+        if (!firstVertex)
+            worldRenderer.endVertex()
+        worldRenderer.pos(Renderer.matrixStack, x, y, z)
+        firstVertex = false
+    }
+
+    /**
+     * Sets the texture location on the last defined vertex.
+     * Use directly after using [Tessellator.pos]
+     *
+     * @param u the u position in the texture
+     * @param v the v position in the texture
+     * @return the Tessellator to allow for method chaining
+     */
+    @JvmStatic
+    fun tex(u: Float, v: Float) = apply {
+        worldRenderer.tex(u.toDouble(), v.toDouble())
+    }
+
+    @JvmStatic
+    @JvmOverloads
+    fun color(x: Float, y: Float, z: Float, a: Float = 1f) = apply {
+        worldRenderer.color(x, y, z, a)
+    }
+
+    @JvmStatic
+    @JvmOverloads
+    fun color(x: Int, y: Int, z: Int, a: Int = 255) = apply {
+        worldRenderer.color(x, y, z, a)
+    }
+
+    @JvmStatic
+    fun color(color: Long) = apply {
+        val (r, g, b, a) = Color(color.toInt())
+        color(r, g, b, a)
+    }
+
+    @JvmStatic
+    fun normal(x: Float, y: Float, z: Float) = apply {
+        worldRenderer.norm(Renderer.matrixStack, x, y, z)
+    }
+
+    @JvmStatic
+    fun overlay(u: Int, v: Int) = apply {
+        worldRenderer.overlay(u, v)
+    }
+
+    @JvmStatic
+    fun light(u: Int, v: Int) = apply {
+        worldRenderer.light(u, v)
+    }
```

Comment:
It's not so clear what these do, so a docstring with a short blurb about their functions might be nice